File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/Drivetrain.java
package org.firstinspires.ftc.teamcode.utils;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

public class Drivetrain
{
    // initialize motors, hardwareMap
    private DcMotor fl, fr, bl, br;
    private HardwareMap hw;

    //constants
    static final double     COUNTS_PER_MOTOR_REV    = 1 ;
    static final double     DRIVE_GEAR_REDUCTION    = 1.0 ;     // No External Gearing.
    static final double     WHEEL_DIAMETER_INCHES   = 4.0 ;     // For figuring circumference
    public static final double     COUNTS_PER_INCH         = (COUNTS_PER_MOTOR_REV * DRIVE_GEAR_REDUCTION) /
            (WHEEL_DIAMETER_INCHES * 3.1415);
    public static final double DRIVE_SPEED = 0.6;
    public static final double TURN_SPEED = 0.5;

    private double drivetrainPowMod = 0.5;


    public Drivetrain (HardwareMap hardwaremap) {
        hw = hardwaremap;

        initialize();
    }

    public void initialize() {
        fr = hw.get(DcMotor.class, "fr");
        fl = hw.get(DcMotor.class, "fl");
        br = hw.get(DcMotor.class, "br");
        bl = hw.get(DcMotor.class, "bl");

        fr.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        fl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        br.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        bl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
        br.setDirection(DcMotorSimple.Direction.REVERSE);
    }

    public void floatMotors() {

        fr.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
        fl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
        br.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
        bl.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT);
    }
    /**
     * Moves drivetrain
     *
     * @param strafe speed of strafe between 0 and 1
     * @param turn speed of turn between 0 and 1
     * @param forward speed of forward between 0 and 1
     */
    public void move(double strafe, double turn, double forward)
    {
        turn *= -drivetrainPowMod;
        forward *= drivetrainPowMod;
        strafe *= -drivetrainPowMod;
        fl.setPower(forward + turn + strafe);
        fr.setPower(forward - turn - strafe);
        bl.setPower(forward + turn - strafe);
        br.setPower(forward - turn + strafe);
        stop();
    }

    public void setMotorPowers(double flpow, double frpow, double blpow, double brpow){
        fl.setPower(flpow);
        fr.setPower(frpow);
        bl.setPower(blpow);
        br.setPower(brpow);
    }
    public void stop() {
        fl.setPower(0);
        fr.setPower(0);
        bl.setPower(0);
        br.setPower(0);

    }

    public void stop_reset_encoder(){
        fl.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        fr.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        bl.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        br.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    }

    public void run_to_pos(){
        fl.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        fr.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        bl.setMode(DcMotor.RunMode.RUN_TO_POSITION);
        br.setMode(DcMotor.RunMode.RUN_TO_POSITION);
    }
    public void no_encoder(){
        fl.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        fr.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        bl.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        br.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    }

    public void run_using_encoder(){
        fl.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        fr.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        bl.setMode(DcMotor.RunMode.RUN_USING_ENCODER);
        br.setMode(DcMotor.RunMode.RUN_TO_POSITION);
    }

    public void setDrivetrainSpeedMod(double speed) {
        drivetrainPowMod=speed;
    }

    public int[] getMotorPositions()
    {
        return new int[]
                {fl.getCurrentPosition(), fr.getCurrentPosition(), bl.getCurrentPosition(), br.getCurrentPosition()};
    }

    public void setPosition(int fl, int fr, int bl, int br){
        this.fl.setTargetPosition(fl);
        this.fr.setTargetPosition(fr);
        this.bl.setTargetPosition(bl);
        this.br.setTargetPosition(br);
    }

    public String printMotorPowers(){
        return "fl: " + fl.getPower() +"fr: " + fr.getPower()+"bl: " + bl.getPower()+"br: " + br.getPower();
    }

    public void forwards(boolean backwards, int LTarget, int RTarget, double speed_mod)
    {
        double pow = drivetrainPowMod * speed_mod;
        int mult = 1;
        if(!backwards) {
            //backwards
            mult *= -1;
        }
        int FLTarget = mult * (int) (LTarget * 31);
        int FRTarget = mult * (int) (RTarget * 31);
        int BLTarget = mult  *(int) (LTarget * 31);
        int BRTarget = mult * (int) (RTarget * 31);


        setPosition(FLTarget, FRTarget, BLTarget, BRTarget);

        run_to_pos();

        setMotorPowers(pow, pow, pow, pow);
    }

    public void strafe(boolean left, int LTarget, int RTarget)
    {
        double pow= drivetrainPowMod;
        int mult = 1;
        if (!left) {
            //right
            mult =-1;
        }

        int FLTarget = mult * (int) (LTarget * 31);
        int FRTarget = -mult * (int) (RTarget * 31);
        int BLTarget = -mult *(int) (LTarget * 31);
        int BRTarget = mult * (int) (RTarget * 31);

        setPosition(FLTarget, FRTarget, BLTarget, BRTarget);

        run_to_pos();

        setMotorPowers(pow, pow, pow,pow);

    }

    public void turn(boolean left, int amount) {
        if (!left) {
            fr.setTargetPosition(-amount);
            fl.setTargetPosition(amount);
            br.setTargetPosition(-amount);
            bl.setTargetPosition(amount);
        } else {
            fr.setTargetPosition(amount);
            fl.setTargetPosition(-amount);
            br.setTargetPosition(amount);
            bl.setTargetPosition(-amount);
        }

        run_to_pos();
    }

    public boolean isBusy(){
        return fl.isBusy() && fr.isBusy() && bl.isBusy() && br.isBusy();
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/hardware/Drivetrain.kt
package org.firstinspires.ftc.teamcode.utils.hardware

import com.qualcomm.robotcore.hardware.*
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit
import java.lang.Math.cos
import java.lang.Math.sin

/**
 * Controls the drivetrain of the robot.
 *
 * This class is designed to control a Mecanum drivetrain using four [DcMotorEx] objects. The class requires that the motors be named frontLeft, frontRight, rearLeft, and rearRight accordingly in the configuration of the robot.
 *
 * @author James Xiao
 * @param hardwareMap The hardware map of the robot.
 * @constructor Creates a drivetrain object using 4 [DcMotorEx] objects from the given [HardwareMap].
 */
class Drivetrain (hardwareMap:HardwareMap) {

    private val frontLeft: DcMotorEx = hardwareMap.get("fl") as DcMotorEx //control hub: 0
    private val frontRight: DcMotorEx = hardwareMap.get("fr") as DcMotorEx //control hub: 1
    private val rearLeft: DcMotorEx = hardwareMap.get("bl") as DcMotorEx //control hub: 2
    private val rearRight: DcMotorEx = hardwareMap.get("br") as DcMotorEx //control hub: 3

    private val motors = listOf(frontLeft, frontRight, rearLeft, rearRight)

    private val imu: IMU = hardwareMap.get("imu") as IMU

    private var powerMultiplier = 1.0
    private var previousPower = 0.0
    private val maxAcceleration = 0.02 // Adjust this value to change the maximum acceleration


    init {
        frontRight.direction = DcMotorSimple.Direction.REVERSE
        rearLeft.direction = DcMotorSimple.Direction.REVERSE
        rearRight.direction = DcMotorSimple.Direction.REVERSE
        frontLeft.direction = DcMotorSimple.Direction.FORWARD

        setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER)
        setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE)
    }

    /**
     * Moves the drivetrain forward and back using [forward], strafes left and right using [strafe],
     * and turns using [turn]. These parameters should be doubles between -1 and 1 representing a proportion
     * of the full power of the motor.
     */
    fun move(forward: Double, strafe: Double, turn: Double) {
        setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER)

        frontLeft.power = powerMultiplier * (forward + turn + strafe)
        frontRight.power = powerMultiplier * (forward - turn - strafe)
        rearLeft.power = powerMultiplier * (forward + turn - strafe)
        rearRight.power = powerMultiplier * (forward - turn + strafe)
    }

    fun moveCentered(forwardBack: Double, rightLeft: Double, rotate: Double) {
        setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER)

        // Get the current heading of the robot
        val heading = getHeading()

        // Calculate the rotated forwardBack and rightLeft values
        val rotatedForwardBack = forwardBack * cos(heading) - rightLeft * sin(heading)
        val rotatedRightLeft = forwardBack * sin(heading) + rightLeft * cos(heading)

        // Calculate the desired power
        val desiredPower = powerMultiplier * (rotatedForwardBack + rotate + rotatedRightLeft)

        // Calculate the change in power
        val powerChange = desiredPower - previousPower

        // Limit the change in power to maxAcceleration
        val limitedPowerChange = when {
            powerChange > maxAcceleration -> maxAcceleration
            powerChange < -maxAcceleration -> -maxAcceleration
            else -> powerChange
        }

        // Calculate the new power
        val newPower = previousPower + limitedPowerChange

        // Set the power of the motors
        frontLeft.power = newPower
        frontRight.power = newPower
        rearLeft.power = newPower
        rearRight.power = newPower

        // Update previousPower
        previousPower = newPower
    }

    /**
     * Moves the drivetrain forward or back [ticks] amount of ticks.
     */
    fun forward(ticks: Int) {
        setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE)

        frontLeft.targetPosition += ticks
        frontRight.targetPosition += ticks
        rearLeft.targetPosition += ticks
        rearRight.targetPosition += ticks

        setMode(DcMotor.RunMode.RUN_TO_POSITION)
    }

    /**
     * Moves the drivetrain left or right [ticks] amount of ticks.
     */
    fun strafe(ticks: Int) {
        setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE)

        frontLeft.targetPosition += ticks
        frontRight.targetPosition -= ticks
        rearLeft.targetPosition -= ticks
        rearRight.targetPosition += ticks

        setMode(DcMotor.RunMode.RUN_TO_POSITION)
    }

    /**
     * Turns the drivetrain left or right [ticks] amount of ticks.
     */
    fun turn(ticks: Int) {
        setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE)

        frontLeft.targetPosition += ticks
        frontRight.targetPosition -= ticks
        rearLeft.targetPosition += ticks
        rearRight.targetPosition -= ticks

        setMode(DcMotor.RunMode.RUN_TO_POSITION)
    }

    /**
     * Floats the drivetrain by resetting the encoders, setting motor power to zero, and make motors
     * float. The function also makes the motors run using the encoders in order to track the
     * movement.
     */
    fun float() {
        setMotorPower(0.0)
        setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER)
        setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.FLOAT)
        setMode(DcMotor.RunMode.RUN_USING_ENCODER)
    }

    /**
     * Sets all of the motor powers to [power]
     */
    fun setMotorPower(power:Double) {
        setMotorPower(power, power, power, power)
    }

    /**
     * Sets each of the motor power values to a given value.
     *
     * @param frontLeftPower The power value for the front left motor.
     * @param frontRightPower The power value for the front right motor.
     * @param rearLeftPower The power value for the rear left motor.
     * @param rearRightPower The power value for the rear right motor.
     *
     */
    fun setMotorPower(frontLeftPower: Double, frontRightPower: Double, rearLeftPower: Double, rearRightPower: Double) {
        frontLeft.power = frontLeftPower
        frontRight.power = frontRightPower
        rearLeft.power = rearLeftPower
        rearRight.power = rearRightPower
    }

    /**
     * Sets the velocity of all motors to [velocity] in ticks per second.
     */
    fun setMotorVelocity(velocity: Double) {
        setMotorVelocity(velocity, velocity, velocity, velocity)
    }

    /**
     * Sets the velocity of all motors to [velocity] in [angleUnit].
     */
    fun setMotorVelocity(velocity: Double, angleUnit: AngleUnit) {
        setMotorVelocity(velocity, velocity, velocity, velocity, angleUnit)
    }

    /**
     * Sets the target velocity of each of the motors in ticks per second.
     *
     * @param frontLeftVelocity The velocity of the front left motor.
     * @param frontRightVelocity The velocity of the front right motor.
     * @param rearLeftVelocity The velocity of the rear left motor.
     * @param rearRightVelocity The velocity of the rear right motor.
     */
    fun setMotorVelocity(frontLeftVelocity: Double, frontRightVelocity: Double, rearLeftVelocity: Double, rearRightVelocity: Double) {
        setMode(DcMotor.RunMode.RUN_USING_ENCODER)

        frontLeft.velocity = frontLeftVelocity
        frontRight.velocity = frontRightVelocity
        rearLeft.velocity = rearLeftVelocity
        rearRight.velocity = rearRightVelocity
    }

    /**
     * Sets the target velocity of each of the motors in the given [AngleUnit].
     *
     * @param frontLeftVelocity The velocity of the front left motor.
     * @param frontRightVelocity The velocity of the front right motor.
     * @param rearLeftVelocity The velocity of the rear left motor.
     * @param rearRightVelocity The velocity of the rear right motor.
     * @param angleUnit The [AngleUnit].
     */
    fun setMotorVelocity(frontLeftVelocity: Double, frontRightVelocity: Double, rearLeftVelocity: Double, rearRightVelocity: Double, angleUnit: AngleUnit) {
        setMode(DcMotor.RunMode.RUN_USING_ENCODER)

        frontLeft.setVelocity(frontLeftVelocity, angleUnit)
        frontRight.setVelocity(frontRightVelocity, angleUnit)
        rearLeft.setVelocity(rearLeftVelocity, angleUnit)
        rearRight.setVelocity(rearRightVelocity, angleUnit)
    }

    /**
     * Uses [behavior] to determine the [DcMotor.ZeroPowerBehavior] of the motors when the power is set to 0.
     */
    fun setZeroPowerBehavior(behavior: DcMotor.ZeroPowerBehavior) {
        applyToAllMotors { it.zeroPowerBehavior = behavior }
    }

    /**
     * Sets the mode of the motors to [DcMotor.RunMode] [mode]
     */
    fun setMode(mode: DcMotor.RunMode) {
        applyToAllMotors { it.mode = mode }
    }

    /**
     * Sets the [powerMultiplier] of the drivetrain to [multiplier]. This value should be a double between -1 and 1.
     */
    fun setPowerMultiplier(multiplier: Double) {
        powerMultiplier = multiplier
    }

    /**
     * Returns the [powerMultiplier] of the drivetrain.
     */
    fun getPowerMultiplier(): Double {
        return powerMultiplier
    }

    /**
     * Returns the [DcMotor.RunMode] of all of the motors if they are all equal, otherwise it will return null.
     */
    fun getMode() = if(listOf(frontLeft.mode, frontRight.mode, rearLeft.mode, rearRight.mode).all{it == frontLeft.mode})frontLeft.mode else null

    /**
     * Returns the [DcMotor.ZeroPowerBehavior] ]of all of the motors if they are all equal, otherwise it will return null.
     */
    fun getZeroPowerBehavior() = if(listOf(frontLeft.zeroPowerBehavior, frontRight.zeroPowerBehavior, rearLeft.zeroPowerBehavior, rearRight.zeroPowerBehavior).all{it == frontLeft.zeroPowerBehavior})frontLeft.zeroPowerBehavior else null

    /**
     * Returns an [Array] of encoder positions for all of the motors in the order front left, front right, rear left, rear right.
     */
    fun getEncoderPositions(): Array<Int> = arrayOf(
            frontLeft.currentPosition,
            frontRight.currentPosition,
            rearLeft.currentPosition,
            rearRight.currentPosition)

    /**
     *  Returns an [Array] containing all of the currents of all of the motors in the drivetrain in [CurrentUnit.AMPS]. It is returned in the order, front left, front right, rear left, rear right
     */
    fun getCurrent(): Array<Double> = arrayOf(
            frontLeft.getCurrent(CurrentUnit.AMPS),
            frontRight.getCurrent(CurrentUnit.AMPS),
            rearLeft.getCurrent(CurrentUnit.AMPS),
            rearRight.getCurrent(CurrentUnit.AMPS)
    )
    fun getAvgCurrent(): Double = getCurrent().average()

    /**
     *  Returns an [Array] containing all of the velocities of all of the motors in the drivetrain in ticks per second. It is returned in the order, front left, front right, rear left, rear right.
     */
    fun getMotorVelocity(): Array<Double> = arrayOf(frontLeft.velocity, frontRight.velocity, rearLeft.velocity, rearRight.velocity)

    /**
     *  Returns an [Array] containing all of the velocities of all of the motors in the drivetrain in the [AngleUnit] [angleUnit]. It is returned in the order, front left, front right, rear left, rear right.
     */
    fun getMotorVelocity(angleUnit: AngleUnit): Array<Double> = arrayOf(frontLeft.getVelocity(angleUnit), frontRight.getVelocity(angleUnit), rearLeft.getVelocity(angleUnit), rearRight.getVelocity(angleUnit))
    fun getMotorPower(): Array<Double> = arrayOf(frontLeft.power, frontRight.power, rearLeft.power, rearRight.power)


    fun getHeading(): Double = imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS)

    /**
     *  Returns a boolean indicating if any of the motors are busy.
     */
    fun isBusy(): Boolean = frontLeft.isBusy || frontRight.isBusy || rearLeft.isBusy || rearRight.isBusy

    private fun applyToAllMotors(action: (DcMotorEx) -> Unit) {
        motors.forEach(action)
    }

    /**
     *  Returns a string containing the power, position, current, and velocity of all of the motors in the drivetrain.
     */

    override fun toString(): String = "frontLeft - power: ${frontLeft.power}, pos: ${frontLeft.currentPosition}, current: ${frontLeft.getCurrent(CurrentUnit.AMPS)}, velocity: ${frontLeft.velocity}\n" +
            "frontRight - power: ${frontRight.power}, pos: ${frontRight.currentPosition}, current: ${frontRight.getCurrent(CurrentUnit.AMPS)}, velocity: ${frontRight.velocity}\n" +
            "rearLeft - power: ${rearLeft.power}, pos: ${rearLeft.currentPosition}, current: ${rearLeft.getCurrent(CurrentUnit.AMPS)}, velocity: ${rearLeft.velocity}\n"+
            "rearRight - power: ${rearRight.power}, pos: ${rearRight.currentPosition}, current: ${rearRight.getCurrent(CurrentUnit.AMPS)}, velocity: ${rearRight.velocity}"
}